contributors/design-proposals/storage/container-storage-interface.md

```diff
@@ -0,0 +1,460 @@
+* Recommend packaging requirements for Kubernetes compatible, third-party CSI Volume drivers.
+* Recommend deployment process for Kubernetes compatible, third-party CSI Volume drivers on a Kubernetes cluster.
+
+## Non-Goals
+* Replace [Flex Volume plugin](https://github.com/kubernetes/community/blob/master/contributors/devel/flexvolume.md)
+  * The Flex volume plugin exists as an exec based mechanism to create “out-of-tree” volume plugins.
+  * Because Flex drivers exist and depend on the Flex interface, it will continue to be supported with a stable API.
+  * The CSI Volume plugin will co-exist with Flex volume plugin.
+
+## Design Overview
+
+To support CSI Compliant Volume plugins, a new in-tree CSI Volume plugin will be introduced in Kubernetes. This new volume plugin will be the mechanism by which Kubernetes users (application developers and cluster admins) interact with external CSI volume drivers.
+
+The `SetUp`/`TearDown` calls for the new in-tree CSI volume plugin will directly invoke `NodePublishVolume` and `NodeUnpublishVolume` CSI RPCs through a unix domain socket on the node machine.
+
+Provision/delete and attach/detach must be handled by some external component that monitors the Kubernetes API on behalf of a CSI volume driver and invokes the appropriate CSI RPCs against it.
+
+To simplify integration, the Kubernetes team will offer a container that captures all the Kubernetes specific logic and acts as a bridge between third-party containerized CSI volume driver and Kubernetes (each deployment of a volume plugin would have it’s own “bridge”).
+
+## Design Details
+
+### Third-Party CSI Volume Drivers
+
+Kubernetes is as minimally prescriptive on the packaging and deployment of a CSI Volume Driver as possible. Use of the *Communication Channels* (documented below) is the only requirement for enabling an arbitrary external CSI compatible storage driver in Kubernetes.
+
+This document recommends a standard mechanism for deploying an arbitrary containerized CSI driver on Kubernetes to simplify deployments of containerized CSI compatible volume drivers on Kubernetes (see the “Recommended Mechanism for Deploying CSI Drivers on Kubernetes” section below), however, this is strictly optional.
+
+### Communication Channels
+
+#### Kubelet to CSI Driver Communication
+
+Kubelet (responsible for mount and unmount) will communicate with an external “CSI volume driver” running on the same host machine (whether containerized or not) via a Unix Domain Socket.
+
+The Unix Domain Socket will be registered with kubelet using the [Device Plugin Unix Domain Socket Registration](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/resource-management/device-plugin.md#unix-socket) mechanism. This mechanism will need to be extended to support registration of both CSI volume drivers and device plugins independently.
```

Comment:
Rewording to:

```
CSI volume drivers should create a socket at the following path on the node machine: `/var/lib/kubelet/plugins/csi/sockets/[driverName]/kubeletproxy.sock`. For alpha, kubelet will assume this is the location for the Unix Domain Socket to talk to the CSI volume driver. For the beta implementation, we can consider using the [Device Plugin Unix Domain Socket Registration](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/resource-management/device-plugin.md#unix-socket) mechanism to register the Unix Domain Socket with kubelet. This mechanism would need to be extended to support registration of both CSI volume drivers and device plugins independently.
```

Reply:
My point was that this registration mechanism seems baroque.  Why should plugins do anything more than create their UNIX socket and wait for kubelet to probe?  That seems simpler and more amenable to multiple driver families.

Reply:
The registration mechanism is the way it is currently implemented by the device plugin. I'm open to your proposal, but I'll leave defer to @jiayingz.

Reply:
Sorry @thockin  I missed your original comment.

It is a good point that we should try to use one-way communication model for its simplicity. We did discuss the option to have kubelet detect new device plugins through inotify during the device plugin design discussion. The main reasons we chose to export a Registration rpc service from Kubelet was mostly because a) it is easier to detect api version mismatch early on during registration, b) we discussed the possible need to authenticate device plugin during registration and grpc has more flexible authentication support, and c) it allows more flexibility for device plugin to choose their communication channel. They don't have to use a unix socket under a canonical path.

Having said these, the current device plugin implementation haven't really used the listed flexibility. The implementation expects a device plugin to export its rpc interface through a socket under the  /var/lib/kubelet/device-plugins directory. Our current security model is that /var/lib/kubelet/device-plugins requires privilege access so a device plugin needs to run with privilege to register with kubelet. From our experience, requiring device plugin to register with kubelet does complicate failure recovery. In particular, when kubelet restarts, there is a time gap for previously registered device plugins to re-register and we have to add some code complexity to populate device plugin resource based on previously checkpointed state and gracefully garbage collect resources from device plugins that fail to re-register. So there are some tradeoffs to consider and the discussions we have here will be very useful to us. Whether we decide to use a poll model or registration model for CSI, we would be interested in following the same model in device plugin to have a consistent communication mechanism across different plugins.